### PR TITLE
notifications: create notification dispatcher

### PR DESCRIPTION
**Pull request by rerowep**
* Adds notification dispatchers

   Try to create a notification and verify received email.

### CODE REVIEW

File: rero_ils/modules/loans/api.py [hunk 1]
```diff
@@ -174,9 +174,12 @@ def create_notification(self, notification_type=None):
                 record['reminder_counter'] = 1
                 notification_to_create = True
         if notification_to_create:
-            data = Notification.create(
+            notification = Notification.create(
                 data=record, dbcommit=True, reindex=True)
-        return data
+            print('====>', notification)
```
Comment (iGormilhit) [inline rero_ils/modules/loans/api.py]: Should this line be removed?

File: rero_ils/modules/loans/api.py [hunk 2]
```diff
@@ -174,9 +174,12 @@ def create_notification(self, notification_type=None):
...
+            notification = notification.dispatch()
+            print('++++>', notification)
```
Comment (iGormilhit) [inline rero_ils/modules/loans/api.py]: And this one too?

File: rero_ils/modules/notifications/dispatcher.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+class Dispatcher():
+    """Dispatcher notifications class."""
+
+    def send(self, notification=None):
+        """Send dispatch."""
+        if notification:
+            data = notification.dumps_pids()
+            self.send_mail(data=data)
+            # notification = notification.update_process_date()
```
Comment (iGormilhit) [inline rero_ils/modules/notifications/dispatcher.py]: Is this line still needed?

File: rero_ils/modules/notifications/dispatcher.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        return notification
+
+    def send_mail(self, data):
+        """Send email."""
+        notification_type = data.get('notification_type')
+        language = 'eng'
```
Comment (iGormilhit) [inline rero_ils/modules/notifications/dispatcher.py]: I don't understand why the language is hard-coded this way.

File: rero_ils/modules/notifications/tasks.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of RERO ILS.
+# Copyright (C) 2017 RERO.
```
Comment (iGormilhit) [inline rero_ils/modules/notifications/tasks.py]: Sometimes it's 2018, sometimes 2017, shouldn't it be 2019?

File: rero_ils/modules/notifications/dispatcher.py (hunk 4 above)
Comment (rerowep) [reply rero_ils/modules/notifications/dispatcher.py]: The language selection for the templates will be resolved in US703

File: rero_ils/modules/items/api_views.py [hunk 6]
```diff
@@ -119,7 +119,8 @@ def librarian_request(item, data):
 
     required_parameters: item_pid_value, location
     """
-    return item.request(**data)
+    request = item.request(**data)
```
Comment (jma) [inline rero_ils/modules/items/api_views.py]: why?

File: rero_ils/modules/loans/api.py [hunk 7]
```diff
@@ -143,7 +143,7 @@ def is_notified(self, notification_type=None):
 
     def create_notification(self, notification_type=None):
         """Creates a recall notification from a checked-out loan."""
-        data = {}
+        notification = {}
```
Comment (jma) [inline rero_ils/modules/loans/api.py]: Why not `notification = None`?

File: rero_ils/modules/notifications/api.py [hunk 8]
```diff
@@ -67,33 +71,130 @@ class Notification(IlsRecord):
...
+                data['loan']['pickup_location'] = \
+                    pickup_location.replace_refs().dumps()
+                # del(data['loan']['pickup_location_pid'])
+            data['loan']['document'] = self.document.replace_refs().dumps()
+            # del(data['loan']['document_pid'])
+            return data
+        except Exception as e:
+            raise(e)
+
+    def set_loan(self):
```
Comment (jma) [inline rero_ils/modules/notifications/api.py]: This name suggest to have a loan as parameter. Perhaps you need to change it.

File: rero_ils/modules/notifications/api.py [hunk 9]
```diff
@@ -67,33 +71,130 @@ class Notification(IlsRecord):
...
+        self = Dispatcher().send(notification=self)
+        return self
+
+    def update_process_date(self):
+        """Update process date."""
+        self['process_date'] = datetime.now().isoformat()
+        self.update(data=self.dumps(), dbcommit=True, reindex=True)
+        return self
+
+    def dumps_pids(self):
```
Comment (jma) [inline rero_ils/modules/notifications/api.py]: Can you gives more informations in the documentations. And perhaps change the name: it suggest to dump a list of pids. Something like: resolve_refs or dumps_with_refs_resolution etc.

File: rero_ils/modules/notifications/dispatcher.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+from flask_security.utils import config_value
+from invenio_mail.api import TemplatedMessage
+from invenio_mail.tasks import send_email
+
+
+class Dispatcher():
+    """Dispatcher notifications class."""
+
+    def send(self, notification=None):
+        """Send dispatch."""
```
Comment (jma) [inline rero_ils/modules/notifications/dispatcher.py]: change it with something like: Send all notifications or Dispatch all notifications, etc.

File: rero_ils/utils.py [hunk 11]
```diff
@@ -42,15 +43,18 @@ def send_mail(subject, recipients, template, language, **context):
...
+                language=language
+            ),
+            **context
+        )
     # if config_value('EMAIL_HTML'):
     #     msg.html = render_template('%s/%s.html' % ctx, **context)
 
-    mail = current_app.extensions.get('mail')
-    mail.send(msg)
+    send_mail_async.delay(msg.__dict__)
```
Comment (jma) [inline rero_ils/utils.py]: can you use `invenio-mail` as in the Dispatcher?

File: tests/conftest.py [hunk 12]
```diff
@@ -25,25 +25,9 @@
 """Common pytest fixtures and plugins."""
 
 import json
-from copy import deepcopy
 from os.path import dirname, join
 
-import mock
 import pytest
-from invenio_circulation.proxies import current_circulation
```
Comment (jma) [inline tests/conftest.py]: I saw it in an other PR. Please check with zub.

File: tests/ui/notifications/test_notifications_api.py [hunk 13]
```diff
@@ -35,9 +38,8 @@
 
 
 def test_notification_create_and_es_mapping(
```
Comment (jma) [inline tests/ui/notifications/test_notifications_api.py]: Can you split this test in two parts: one for the mapping and an other one for the creation?

File: tests/ui/notifications/test_notifications_api.py [hunk 14]
```diff
@@ -64,6 +67,10 @@ def test_notification_create_and_es_mapping(
     assert fetched_pid.pid_value == pid
     assert fetched_pid.pid_type == 'notif'
 
+    notification.dispatch()
+    # print('--->', mailbox)
```
Comment (jma) [inline tests/ui/notifications/test_notifications_api.py]: please remove this

File: rero_ils/modules/items/api_views.py (hunk 6 above)
Comment (rerowep) [reply rero_ils/modules/items/api_views.py]: Corrected, forgot after debugging.

File: rero_ils/modules/loans/api.py (hunk 7 above)
Comment (rerowep) [reply rero_ils/modules/loans/api.py]: This code is from Aly, I dont know.

File: rero_ils/modules/notifications/api.py (hunk 8 above)
Comment (rerowep) [reply rero_ils/modules/notifications/api.py]: Changed to `init_loan`

File: rero_ils/modules/notifications/api.py (hunk 9 above)
Comment (rerowep) [reply rero_ils/modules/notifications/api.py]: Changed to `replace_pids_and_refs`

File: rero_ils/utils.py (hunk 11 above)
Comment (rerowep) [reply rero_ils/utils.py]: This is the `send_mail` from `invenio-mail`.

File: rero_ils/modules/notifications/dispatcher.py (hunk 10 above)
Comment (rerowep) [reply rero_ils/modules/notifications/dispatcher.py]: Changed to: `dispatch_notification`

File: tests/conftest.py (hunk 12 above)
Comment (rerowep) [reply tests/conftest.py]: `current_circulation` is not used in `tests/conftest.py`

File: tests/ui/notifications/test_notifications_api.py (hunk 13 above)
Comment (rerowep) [reply tests/ui/notifications/test_notifications_api.py]: This code is from Aly, have to ask him!

**Comment by BadrAly** [comment]
Emails are nicely (verified the gmail inbox) sent after the creation of the `availability` and `recall` notifications.

File: rero_ils/modules/loans/api.py (hunk 7 above)
Comment (BadrAly) [reply rero_ils/modules/loans/api.py]: Cause we are returning a dict.  I can change it to None if necessary.

File: tests/ui/notifications/test_notifications_api.py (hunk 13 above)
Comment (BadrAly) [reply tests/ui/notifications/test_notifications_api.py]: Merged these two functions in one to avoid duplicating the test of the notification creation.

**Pull request merged**